### ISSUE TITLE
Version 6.12.0 contains regression for resolving modules/actions.

### ISSUE THREAD

**Issue by TylerJang27** [ansible/ansible-lint#2969]
##### Summary

Ansible-lint fails to resolve some modules/actions in playbooks.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
ansible-lint --version  # 6.12.0
```

- ansible-lint installation method: pip installed (using [trunk](https://docs.trunk.io/docs/check)) on Ubuntu 20.04.5 LTS

##### STEPS TO REPRODUCE

```yaml
# jboss-standalone/demo-aws-launch.yml (relative to workspace)
---
- name: Provision instances
  hosts: localhost
  connection: local
  gather_facts: False

  # load AWS variables from this group vars file
  vars_files:
    - group_vars/all

  tasks:
    - name: Launch instances
      ec2:
        access_key: "{{ ec2_access_key }}"
        secret_key: "{{ ec2_secret_key }}"
        keypair: "{{ ec2_keypair }}"
        group: "{{ ec2_security_group }}"
        type: "{{ ec2_instance_type }}"
        image: "{{ ec2_image }}"
        region: "{{ ec2_region }}"
        instance_tags:
          "{'ansible_group':'jboss', 'type':'{{ ec2_instance_type }}', 'group':'{{
          ec2_security_group }}', 'Name':'demo_''{{ tower_user_name }}'}"
        count: "{{ ec2_instance_count }}"
        wait: true
      register: ec2

    - name: Wait for SSH to come up
      wait_for:
        host: "{{ item.public_dns_name }}"
        port: 22
        delay: 60
        timeout: 320
        state: started
      with_items: "{{ ec2.instances }}"

```

```console (paste below)
ansible-lint
```

##### Desired Behavior

< 6.12.0 (specifically 6.11.0):
```
yaml[indentation]: Wrong indentation: expected 4 but found 2
.trunk/trunk.yaml:6

yaml[truthy]: Truthy value should be one of [false, true]
jboss-standalone/demo-aws-launch.yml:5

role-name: Role name ZbSoIn does not match ``^[a-z][a-z0-9_]*$`` pattern. (warning)
tmp/trunk-1001/ZbSoIn:1

You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - yaml[indentation]  # Violations reported by yamllint.
  - yaml[truthy]  # Violations reported by yamllint.

                      Rule Violation Summary                      
 count tag               profile rule associated tags             
     1 role-name         basic   deprecations, metadata (warning) 
     1 yaml[indentation] basic   formatting, yaml                 
     1 yaml[truthy]      basic   formatting, yaml                 

Failed after min profile: 2 failure(s), 1 warning(s) on 14 files.
```
##### Actual Behavior

\>= 6.12.0:
```
syntax-check[specific]: couldn't resolve module/action 'ec2'. This often indicates a misspelling, missing collection, or incorrect module path.
jboss-standalone/demo-aws-launch.yml:12:7

You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - syntax-check[specific]  # Ansible syntax check failed.

                  Rule Violation Summary                   
 count tag                    profile rule associated tags 
     1 syntax-check[specific] min     core                 

Failed after : 1 failure(s), 0 warning(s) on 14 files.
```

Issue was first identified when running [newly released version with trunk](https://github.com/trunk-io/plugins/actions/runs/4068099243/jobs/7006240066)

**Comment by hollow** [comment]
same here – just spent two hours trying to figure out why allall of my ansible projects are failing pre-commit after updating today ...

**Comment by ssbarnea** [comment]
Sorry but this reports sounds more of a support ticket than a bug and I will likely have to convert it to a discussion.

What I see for sure here is that lack of use of fully qualified module names. Read https://ansible-lint.readthedocs.io/rules/fqcn/#correct-code

Also remember that "collections" keyword is not supported.

**Comment by hollow** [comment]
@ssbarnea this bug also exists for FQCNs

**Comment by ssbarnea** [comment]
@hollow In this case, please try to produce a minimal example that reproduces this bug an nothing else. This should help us reproduce it and produce a fix.

While using pre-commit you need to be careful because their SaaS version does not have network access, so it cannot install required collections. If you run it yourself, you do not have this problem. Also, you could use https://github.com/ansible/ansible-lint-action which is the GHA we maintain. Still, with the action there is a slight delay of few days as we do not update it instantly after releasing a new linter, especially if the release was not a patch version.

**Comment by hollow** [comment]
@ssbarnea while trying to come up with a minimal example I have realized that the actual issue seems to be burried in a lot of warnings that ansible-lint 6.12.0+ produces:

```
WARNING  Unable to resolve FQCN for module community.libvirt.virt
WARNING  Unable to resolve FQCN for module netbox.netbox.netbox_tag
WARNING  Unable to resolve FQCN for module netbox.netbox.netbox_virtual_machine
WARNING  Unable to load module community.general.nomad_job at playbooks/pushgateway.yaml:7 for options validation
WARNING  Unable to resolve FQCN for module community.general.nomad_job
WARNING  Unable to load module netbox.netbox.netbox_device at playbooks/tasks/netbox_tags.yaml:58 for options validation
WARNING  Unable to load module netbox.netbox.netbox_virtual_machine at playbooks/tasks/netbox_tags.yaml:70 for options validation
WARNING  Unable to resolve FQCN for module netbox.netbox.netbox_device
WARNING  Unable to load module community.general.parted at playbooks/tasks/optimize_nvme.yaml:22 for options validation
WARNING  Unable to resolve FQCN for module community.general.parted
```

the list goes on and on, but eventually ansible-lint fails with:

```
jinja[invalid]: An unhandled exception occurred while templating '{{ lookup('community.general.dig', item.key) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: lookup plugin (community.general.dig) not found
playbooks/pxe.yaml:74 Task/Handler: Create kickstart files
```

I have extracted the affected task and reproduced the error with the ansible-lint action without pre-commit here: https://github.com/hollow/ansible-lint-regression/actions/runs/4083566375/jobs/7039250446

Not sure if the issue is related to all the new warnings about not being able to load modules though.

**Comment by eff917** [comment]
I have managed to reproduce this issue with the following minimal setup:
 create a single file in a role:
`roles/relative_regression/tasks/main.yml`
```---
- name: 'Copy configuration files'
  ansible.posix.synchronize:
    src: configs/
    dest: /etc/configs/
```
6.12.0 and 6.12.1 output:
`ansible-lint roles/relative_regression/` output:
```
[WARNING]: errors were encountered during the plugin load for ansible.posix.synchronize: ['relative resource paths
not supported']
WARNING  Unable to load module ansible.posix.synchronize at roles/relative_regression/tasks/main.yml:2 for options validation
WARNING  Unable to resolve FQCN for module ansible.posix.synchronize

Passed with production profile: 0 failure(s), 0 warning(s) on 2 files.
```

6.11.0 output:
```
Passed with production profile: 0 failure(s), 0 warning(s) on 2 files.
A new release of ansible-lint is available: 6.11.0 → 6.12.1 Upgrade by running: pip install --upgrade ansible-lint
```
The import warning only seems to be present inside roles.
Also just noticed it says 2 files, when there's only one file and one directory inside the folder. If I directly call on the file, it correctly shows 1 file scanned.

**Comment by skokhanovskiy** [comment]
@hollow @ssbarnea here is the minimal `Dockerfile` for reproduction of this issue.

```Dockerfile
# syntax=docker/dockerfile:1

FROM debian

RUN apt-get update && apt install -y git python3-pip python3-wheel

RUN pip install ansible-core==2.13.7 && ansible --version
RUN ansible-galaxy collection install community.general:==6.3.0

RUN pip install ansible-lint==6.11.0 && ansible-lint --version

COPY <<-EOF /playbook.yml
---
- name: Test ansible-lint
  hosts: all
  vars:
    pam_rhel_modules: []
  tasks:
    - name: Ensure pam modules are configured
      community.general.pamd:
        foo: bar
        bar: foo
      loop_control:
        label: "{{ item.name }}"
      loop: "{{ pam_rhel_modules }}"
EOF

ENTRYPOINT [ "ansible-lint", "-v", "/playbook.yml" ]

```

Output of the container:

```
WARNING  Ignore loading rule from /usr/local/lib/python3.9/dist-packages/ansiblelint/rules/galaxy.py due to No module named 'pytest'
INFO     Set ANSIBLE_LIBRARY=/root/.cache/ansible-compat/e3b0c4/modules:/root/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/root/.cache/ansible-compat/e3b0c4/collections:/root/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/root/.cache/ansible-compat/e3b0c4/roles:/root/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
INFO     Executing syntax check on playbook.yml (0.44s)
WARNING  Unable to load module community.general.pamd at playbook.yml:7 for options validation
WARNING  Unable to resolve FQCN for module community.general.pamd
WARNING  Listing 1 violation(s) that are fatal
name[casing]: All names should start with an uppercase letter. (warning)
playbook.yml:2


              Rule Violation Summary              
 count tag          profile  rule associated tags 
     1 name[casing] moderate idiom (warning)      

Passed with basic profile, 1/5 star rating: 0 failure(s), 1 warning(s) on 1 files.
```


For previous version ansible-lint everything works fine:

```
INFO     Set ANSIBLE_LIBRARY=/root/.cache/ansible-compat/e3b0c4/modules:/root/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/root/.cache/ansible-compat/e3b0c4/collections:/root/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/root/.cache/ansible-compat/e3b0c4/roles:/root/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
INFO     Executing syntax check on playbook.yml (0.43s)
WARNING  Listing 1 violation(s) that are fatal
args[module]: missing required arguments: control, module_path, name, type (warning)
playbook.yml:7 Task/Handler: Ensure pam modules are configured

You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - experimental  # all rules tagged as experimental

                  Rule Violation Summary                   
 count tag          profile rule associated tags           
     1 args[module]         syntax, experimental (warning) 

Passed with production profile, 5/5 star rating: 0 failure(s), 1 warning(s) on 1 files.
A new release of ansible-lint is available: 6.11.0 → 6.12.1
```

**Comment by skokhanovskiy** [comment]
IMHO this issue connected with closed #2828 that @ssbarnea moved to discussions #2861.

**Comment by kristof-mattei** [comment]
@skokhanovskiy 
> Output of the container:
> 
> ```
> WARNING  Ignore loading rule from /usr/local/lib/python3.9/dist-packages/ansiblelint/rules/galaxy.py due to No module named 'pytest'
> INFO     Set ANSIBLE_LIBRARY=/root/.cache/ansible-compat/e3b0c4/modules:/root/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
> INFO     Set ANSIBLE_COLLECTIONS_PATH=/root/.cache/ansible-compat/e3b0c4/collections:/root/.ansible/collections:/usr/share/ansible/collections
> INFO     Set ANSIBLE_ROLES_PATH=/root/.cache/ansible-compat/e3b0c4/roles:/root/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
> INFO     Executing syntax check on playbook.yml (0.43s)
> WARNING  Unable to load module community.general.pamd at playbook.yml:7 for options validation
> WARNING  Unable to resolve FQCN for module community.general.pamd
> 
> Passed with production profile: 0 failure(s), 0 warning(s) on 1 files.
> ```


`python -m pip install pytest` seems to fix it for me.

**Comment by skokhanovskiy** [comment]
@kristof-mattei In my case installing of the pytest module resolves only the first warning.

Check this out.

```Dockerfile
# syntax=docker/dockerfile:1

FROM debian

RUN apt-get update && apt install -y git python3-pip python3-wheel

RUN pip install ansible-core==2.13.7 && ansible --version
RUN ansible-galaxy collection install community.general:==6.3.0

RUN pip install pytest ansible-lint==6.12.0 && ansible-lint --version

COPY <<-EOF /playbook.yml
---
- name: Test ansible-lint
  hosts: all
  vars:
    pam_rhel_modules: []
  tasks:
    - name: Ensure pam modules are configured
      community.general.pamd:
        foo: bar
        bar: foo
      loop_control:
        label: "{{ item.name }}"
      loop: "{{ pam_rhel_modules }}"
EOF

ENTRYPOINT [ "ansible-lint", "-v", "/playbook.yml" ]
```

```bash
docker build -t test-ansible-lint .
```
```bash
docker run --rm test-ansible-lint
```

We are expecting error message from linter about missed parameters:

```
args[module]: missing required arguments: control, module_path, name, type (warning)
playbook.yml:7 Task/Handler: Ensure pam modules are configured
```

But gets only this:

```
WARNING  Unable to load module community.general.pamd at playbook.yml:7 for options validation
WARNING  Unable to resolve FQCN for module community.general.pamd

Passed with production profile: 0 failure(s), 0 warning(s) on 1 files.
```

**Comment by bluikko** [comment]
@hollow are you using offline mode? I am seeing also a ton of those warnings + errors, it seems everything outside of `ansible.builtin` is not found.

**Comment by hollow** [comment]
@bluikko I'm not using offline mode. Despite the warnings (which I ignore for now) I have mitigated the one remaining error (that I have reproduced in the linked repo above) with a `# noqa` directive 🤷‍♂️

**Comment by apatard** [comment]
hm. After seeing the same issue, I may have an idea of what's going on.
Can you someone try the following workaround please ?

```
--- ansiblelint/_vendor/ansible_compat/runtime.py.orig	2023-02-13 15:42:31.266911773 +0100
+++ ansiblelint/_vendor/ansible_compat/runtime.py	2023-02-13 15:44:51.132685168 +0100
@@ -142,11 +142,10 @@ class Runtime:
 
                 # noinspection PyProtectedMember
                 # pylint: disable=protected-access
-                _AnsibleCollectionFinder(
-                    paths=[
-                        os.path.dirname(os.environ.get(ansible_collections_path(), "."))
-                    ]
-                )._install()  # pylint: disable=protected-access
+                col_path = [f"{self.cache_dir}/collections"]
+                col_path += self.config.collections_paths
+                col_path += [os.path.dirname(os.environ.get(ansible_collections_path(), "."))]
+                _AnsibleCollectionFinder(paths=col_path)._install()  # pylint: disable=protected-access
             Runtime.initialized = True
 
     def clean(self) -> None:
```

**Comment by apatard** [comment]
@skokhanovskiy thanks for testing. I'm working on a PR for ansible-compat. If it gets merged, then the local version needs to be updated.
